ext/sodium/sodium_pwhash.c

```diff
@@ -0,0 +1,192 @@
+}
+
+static zend_bool php_sodium_argon2_needs_rehash(const zend_string *hash, zend_array *options) {
+	size_t opslimit = crypto_pwhash_argon2id_OPSLIMIT_MODERATE;
+	size_t memlimit = crypto_pwhash_argon2id_MEMLIMIT_MODERATE;
+
+	if (options) {
+		zval *opt;
+		if ((opt = zend_hash_str_find(options, "memory_cost", strlen("memory_cost")))) {
+			memlimit = zval_get_long(opt);
+			if ((memlimit < crypto_pwhash_MEMLIMIT_MIN) || (memlimit >crypto_pwhash_MEMLIMIT_MAX)) {
+				php_error_docref(NULL, E_WARNING, "Memory cost is outside of allowed memory range");
+				return 1;
+			}
+		}
+		if ((opt = zend_hash_str_find(options, "time_cost", strlen("time_cost")))) {
+			opslimit = zval_get_long(opt);
+			if ((opslimit < crypto_pwhash_OPSLIMIT_MIN) || (opslimit >crypto_pwhash_OPSLIMIT_MAX)) {
+				php_error_docref(NULL, E_WARNING, "Time cost is outside of allowed time range");
+				return 1;
+			}
+		}
+	}
+
+	return crypto_pwhash_str_needs_rehash(ZSTR_VAL(hash), opslimit, memlimit);
+}
+
+static int php_sodium_argon2_get_info(zval *return_value, const zend_string *hash) {
+	const char* p = NULL;
+	zend_long v = 0, threads = 1;
+	zend_long memory_cost = crypto_pwhash_argon2id_MEMLIMIT_MODERATE;
+	zend_long time_cost = crypto_pwhash_argon2id_OPSLIMIT_MODERATE;
+
+	if (!hash || (ZSTR_LEN(hash) < sizeof("$argon2id$"))) {
+		return FAILURE;
+	}
+
+	p = ZSTR_VAL(hash);
+	if (!memcmp(p, "$argon2i$", strlen("$argon2i$"))) {
+		p += strlen("$argon2i$");
+	} else if (!memcmp(p, "$argon2id$", strlen("$argon2id$"))) {
+		p += strlen("$argon2id$");
+	} else {
+		return FAILURE;
+	}
+
+	sscanf(p, "v=" ZEND_LONG_FMT "$m=" ZEND_LONG_FMT ",t=" ZEND_LONG_FMT ",p=" ZEND_LONG_FMT,
+	       &v, &memory_cost, &time_cost, &threads);
+
+	add_assoc_long(return_value, "memory_cost", memory_cost);
+	add_assoc_long(return_value, "time_cost", time_cost);
+	add_assoc_long(return_value, "threads", threads);
+
+	return SUCCESS;
+}
+
+/* argon2i specific methods */
+
+static zend_string *php_sodium_argon2i_hash(const zend_string *password, zend_array *options) {
+    return php_sodium_argon2_hash(password, options, crypto_pwhash_ALG_ARGON2I13);
+}
+
+static const php_password_algo sodium_algo_argon2i = { 
+    "argon2i",
+    php_sodium_argon2i_hash,
+    php_sodium_argon2_verify,
+    php_sodium_argon2_needs_rehash,
+    php_sodium_argon2_get_info,
+    NULL,
+};
+
+/* argon2id specific methods */
+
+static zend_string *php_sodium_argon2id_hash(const zend_string *password, zend_array *options) {
+    return php_sodium_argon2_hash(password, options, crypto_pwhash_ALG_ARGON2ID13);
+}
+
+static const php_password_algo sodium_algo_argon2id = { 
+    "argon2id",
+    php_sodium_argon2id_hash,
+    php_sodium_argon2_verify,
+    php_sodium_argon2_needs_rehash,
+    php_sodium_argon2_get_info,
+    NULL,
+};
+
+PHP_MINIT_FUNCTION(sodium_password_hash) /* {{{ */ {
+	zend_string *argon2i = zend_string_init("argon2i", strlen("argon2i"), 1);
+
+	if (php_password_algo_find(argon2i)) {
```

Comment:
Would it be simpler to conditionally compile this stuff, `#ifndef HAVE_ARGON2LIB` ?

(Yes someone can build the shared sodium ext, and try to use it on another system with a PHP build with different option... which seems terrible ugly way...)

Reply:
"simpler" is a relative term here.  The algorithm implementations aren't /that/ large as potentially dead code goes.  If someone feels strongly about it, then I don't mind going with an ifdef instead of a runtime check, but my inclination is to lean towards letting runtime be the cononical source of truth, not build info.

Reply:
OK, fine for me.